swarm/pss/protocol.go

```
@@ -0,0 +1,201 @@
+
+// Injects a p2p.Msg into the MsgReadWriter, so that it appears on the associated p2p.MsgReader
+func (prw PssReadWriter) injectMsg(msg p2p.Msg) error {
+	log.Trace(fmt.Sprintf("pssrw injectmsg: %v", msg))
+	prw.rw <- msg
+	return nil
+}
+
+// For devp2p protocol integration only.
+//
+// Convenience object for passing messages in and out of the p2p layer
+type PssProtocol struct {
+	*Pss
+	proto        *p2p.Protocol
+	topic        *whisper.TopicType
+	spec         *protocols.Spec
+	pubKeyRWPool map[string]p2p.MsgReadWriter
+	symKeyRWPool map[string]p2p.MsgReadWriter
+	flags        byte
+}
+
+// For devp2p protocol integration only.
+//
+// Maps a Topic to a devp2p protocol.
+//
+// flags: 0x01 = asymmetric messaging enabled, 0x02 symmetric messaging enabled
+func RegisterPssProtocol(ps *Pss, topic *whisper.TopicType, spec *protocols.Spec, targetprotocol *p2p.Protocol, flags byte) (*PssProtocol, error) {
+	if flags&0x03 == 0 {
+		return nil, fmt.Errorf("specify at least one of asymmetric or symmetric messaging mode")
+	}
+	pp := &PssProtocol{
+		Pss:          ps,
+		proto:        targetprotocol,
+		topic:        topic,
+		spec:         spec,
+		pubKeyRWPool: make(map[string]p2p.MsgReadWriter),
+		symKeyRWPool: make(map[string]p2p.MsgReadWriter),
+		flags:        flags,
+	}
+	return pp, nil
+}
+
+func (self *PssProtocol) isAsymmetric() bool {
+	return self.flags&0x01 != 0
+}
+
+func (self *PssProtocol) isSymmetric() bool {
+	return self.flags&0x02 != 0
+}
+
+// For devp2p protocol integration only.
+//
+// BROKEN! Implementation for pubkey lookup must be implemented
+//
+// Generic handler for initiating devp2p-like protocol connections
+//
+// This handler should be passed to Pss.Register with the associated topic.
+func (self *PssProtocol) Handle(msg []byte, p *p2p.Peer, asymmetric bool, keyid string) error {
+	var vrw *PssReadWriter
+	if self.isAsymmetric() != asymmetric && self.isSymmetric() == !asymmetric {
+		return fmt.Errorf("invalid protocol encryption")
```

Use `errors.New`

---

`grep -ir fmt.err` in the ethereum repo reveals a *lot* of `return fmt.Error`. We should change them all?

---

Well, `fmt.Errorf` is for when you want to "format" an error string (see [here](https://golang.org/pkg/fmt/#Errorf)), whereas if you just want to return an error string then `errors.New` is more appropriate.

As for changing them all, that work is [underway](https://github.com/ethereum/go-ethereum/pull/14972).